wrapper.c

```diff
@@ -4,6 +4,39 @@
 #include "cache.h"
 #include "config.h"
 
+#ifdef USE_MSVC_CRTDBG
+#undef strdup         /* cancel definitions from crtdbg.h because they hard */
+#undef calloc         /* code __FILE__ and __LINE__ */
+#undef malloc
+#undef realloc
+
+#define strdup(s)     (_strdup_dbg((s),       _NORMAL_BLOCK, caller_fn, caller_ln))
```

Comment:
How does this `caller_fn`/`caller_ln` business work? I do not see `strdup()` declaring such parameters...

Reply:
The only reference to "strdup" is inside of xstrdup().  When the DCL_ macros are in effect those args come from the DCL_ macro.  This allows the __FILE__, __LINE__ from the caller of xstrdup() to be passed thru to the CRTDBG routines.

We could get by with out this.  I was just trying to minimize the amount of code changes in the main part of the file.
With what I have, I only had to change the function signatures with the DCL_ macros and not change anything within the functions.

Reply:
Okay, I think I get it now.

Somehow I do not believe that Junio would like these `DCL_*()`.

I could, however, see these patches being accepted if we transformed them in such a way that it was less specific to CRTDBG. So if we would pretend, for example, that the file/line was required for, say, proper tracing...?

Reply:
Yeah, I wasn't sure how to do that.  I was afraid that they wouldn't like the overhead of
passing 2 extra args to every malloc call for production code and this code base doesn't
seem to try to avoid lots of #ifdef's.

If we want to try for non-crtdbg-specific tracing that might make it easier to modify
the macros/prototypes -- I was trying to hold the Linux side as constant as possible.

Perhaps I could put the extra args under an #ifdef DEBUG  or #ifdef DEBUG_LEAKS
and not something platform specific.

Reply:
> #ifdef DEBUG_LEAKS

I like that.

If only we could create a "template" that we could then apply to wrappers around all the `*alloc*()` functions...